File: week_1/credit/credit.c
#include<cs50.h>
#include<stdio.h>
#include<math.h>

int main(void)
{
    // Get card number
    long n = get_long("Number: ");

    // Count length
    int i = log10(n) + 1;

    printf("%i \n", i);

    // Check if length is valid
    if (i != 13 && i != 15 && i != 16)
    {
        printf("INVALID\n");
        return 0;
    }

    // Calculate checksum
    int sum1 = 0;
    int sum2 = 0;
    long x = n;
    int total = 0;
    int mod1;
    int mod2;
    int d1;
    int d2;

    do
    {
        // Remove last digit and add to sum1
        mod1 = x % 10;
        x = x / 10;
        sum1 = sum1 + mod1;

        // Remove second last digit
        mod2 = x % 10;
        x = x / 10;

        // Double second last digit and add digits to sum2
        mod2 = mod2 * 2;
        d1 = mod2 % 10;
        d2 = mod2 / 10;
        sum2 = sum2 + d1 + d2;
    }
    while (x > 0);
    total = sum1 + sum2;



    // Next check Luhn Algorithm
    if (total % 10 != 0)
    {
        printf("INVALID\n");
        return 0;
    }


    // Get starting digits
    long start = n;
    do
    {
        start = start / 10;
    }
    while (start > 100);

    // Next check starting digits for card type
    if ((start / 10 == 5) && (0 < start % 10 && start % 10 < 6))
    {
        printf("MASTERCARD\n");
    }

    else if ((start / 10 == 3) && (start % 10 == 4 || start % 10 == 7))
    {
        printf("AMEX\n");
    }

    else if (start / 10 == 4)
    {
        printf("VISA\n");
    }

    else
    {
        printf("INVALID\n");
    }

}

File: week_6/credit/credit.py
# Statics
AE1 = "34"
AE2 = "37"
MC1 = "51"
MC2 = "52"
MC3 = "53"
MC4 = "54"
MC5 = "55"
Visa = "4"

# Vars
sum1 = 0
sum2 = 0
temp_int = 0

# get input from the user
card_num = input("Number: ")

# Calculating len of input
card_len = len(card_num)
brand = "INVALID"

# Compare first 2 digits to the brands
if (card_len == 13):
    if (card_num[:1] == Visa):
        brand = "VISA"

elif (card_len == 15):
    if (card_num[:2] == AE1 or card_num[:2] == AE2):
        brand = "AMEX"

elif (card_len == 16):
    if (card_num[:2] == MC1 or card_num[:2] == MC2 or card_num[:2] == MC3 or card_num[:2] == MC4 or card_num[:2] == MC5):
        brand = "MASTERCARD"
    elif (card_num[:1] == Visa):
        brand = "VISA"

else:
    print("INVALID")
    # End the program
    exit()


# Change string to int
card_num = int(card_num)

# Checksum input
while(card_num > 0):

    # Remove last digit and add to sum1
    sum1 += card_num % 10
    card_num = card_num // 10

    # Remove second last digit
    temp_int = (card_num % 10)*2
    card_num = card_num // 10

    # If the number is greater than 10 (sum of the first digit and the second digit)
    # For example 12 ==> 1+2 , 3 ==> 3+0
    sum2 += (temp_int % 10)+(temp_int // 10)

# if it validates the Algorithm prints result
if ((sum1 + sum2) % 10 == 0):
    print(brand)

else:
    print("INVALID")